Zd vcr billy

ZmagoD opened the pull request:
1. clean the  stubs in features/support/webmoc.rb and features/step_definition/map_steps.rb ( stub_request_with_address function)
2.  install the vcr gem and make a setup like in the doit_rails app 
 - set the vcr.rb config file in the features/support/ directory
 - add the VCR.use_cassette when creating the organisations in the features/step_definition/factory_steps.rb to the step " Given /^the following organisations exist:$/ do |organisations_table| "
 - added the "@vcr" tag in every scenario that the gem finds that interacts with third party calls



online I'm getting 1 error 
#### cucumber features/home_page/map.feature:65

![screenshot from 2016-01-23 23 36 20](https://cloud.githubusercontent.com/assets/7046787/12533305/2cec9f62-c22c-11e5-9039-87a1554f37a9.png)

offline I get 16 errors mainly related to google maps and the displaying of markers.

ZmagoD commented:
1. Added puffing-billy gem 
2. Add the configuration code like in doit_rails in the features/support/env.rb and spec/rails_helper.rb
3. Added "@billy" tags on the previous failing specs offline
 - the cucumber features/home_page/map.feature:65 still persist an 4 new errors aper ( maybe because they are https url and we have to change something in the billy gem configuration)
![screenshot from 2016-01-24 00 45 59](https://cloud.githubusercontent.com/assets/7046787/12533725/47aea0c6-c23a-11e5-958c-63d8b42924cc.png)

details 
![screenshot from 2016-01-24 01 00 59](https://cloud.githubusercontent.com/assets/7046787/12533729/4f11287a-c23a-11e5-9abb-3241c6d895f1.png)

 - other previously falling test all passes offline and online

tansaku commented:
great work getting started on this @ZmagoD - I'm checking this out now

tansaku commented:
Working through this and wondering if it might be easier if we had blanket VCR coverage of cucumber tests rather than needing to turn it on and off again on a scenario by scenario basis ...?

The danger I see is developers forgetting to add the tag and/or not understanding it ... what are the upsides of tagging on a scenario by scenario basis?  Performance only?

tansaku commented:
Locally and online, running the entire cucumber suite I get 4 failures:

```
(::) failed steps (::)

no implicit conversion of Symbol into Integer (TypeError)
./features/step_definitions/authentication_steps.rb:99:in `/^I have a "([^\"]+)" cookie set to "([^\"]+)"$/'
./features/step_definitions/authentication_steps.rb:103:in `/^cookies are approved$/'
features/home_page/map.feature:69:in `Given cookies are approved'

no implicit conversion of Symbol into Integer (TypeError)
./features/step_definitions/authentication_steps.rb:99:in `/^I have a "([^\"]+)" cookie set to "([^\"]+)"$/'
./features/step_definitions/authentication_steps.rb:103:in `/^cookies are approved$/'
features/volunteer_opportunities/map_volunteer_opportunities.feature:23:in `And cookies are approved'

no implicit conversion of Symbol into Integer (TypeError)
./features/step_definitions/authentication_steps.rb:99:in `/^I have a "([^\"]+)" cookie set to "([^\"]+)"$/'
./features/step_definitions/authentication_steps.rb:103:in `/^cookies are approved$/'
features/volunteer_opportunities/map_volunteer_opportunities.feature:30:in `And cookies are approved'

no implicit conversion of Symbol into Integer (TypeError)
./features/step_definitions/authentication_steps.rb:99:in `/^I have a "([^\"]+)" cookie set to "([^\"]+)"$/'
./features/step_definitions/authentication_steps.rb:103:in `/^cookies are approved$/'
features/volunteer_opportunities/map_volunteer_opportunities.feature:30:in `And cookies are approved'

Failing Scenarios:
cucumber features/home_page/map.feature:66 # Scenario: Changing postcode changes the map coordinates
cucumber features/volunteer_opportunities/map_volunteer_opportunities.feature:21 # Scenario: See a map of current volunteer opportunities
cucumber features/volunteer_opportunities/map_volunteer_opportunities.feature:28 # Scenario: Volunteer opportunites are listed in map popups
cucumber features/volunteer_opportunities/map_volunteer_opportunities.feature:28 # Scenario: Volunteer opportunites are listed in map popups


```

and I also notice that I still get a lot of webmock(?) messages in the log output:

```
...Request to unknown URL: http://csi.gstatic.com/csi?v=2&s=mapsapi3&action=map&rt=firsttile.214
To block requests to unknown URLs:
  page.driver.block_unknown_urls
To allow just this URL:
  page.driver.allow_url("http://csi.gstatic.com/csi?v=2&s=mapsapi3&action=map&rt=firsttile.214")
To allow requests to URLs from this host:
  page.driver.allow_url("csi.gstatic.com")
Request to unknown URL: http://csi.gstatic.com/csi?v=2&s=mapsapi3&action=apiboot&rt=jl.7,mc.37,mb.85,smv.108,dm.108,prt.108,plt.78,ep.235,vt.253,mt.254&size=556x609&maptype=m
To block requests to unknown URLs:
  page.driver.block_unknown_urls
To allow just this URL:
  page.driver.allow_url("http://csi.gstatic.com/csi?v=2&s=mapsapi3&action=apiboot&rt=jl.7,mc.37,mb.85,smv.108,dm.108,prt.108,plt.78,ep.235,vt.253,mt.254&size=556x609&maptype=m")
To allow requests to URLs from this host:
  page.driver.allow_url("csi.gstatic.com")
```

I wonder if this relates to scenarios that are not yet covered by billy?  

I think my approach to investigate all this if I get the chance today is to review each of those four failures and then the scenarios one by one, as well as the failure in the CI ...

ZmagoD commented:
Hi @tansaku, thank you for the review. Yes you are right it will be better to if we set the VCR coverage for all cucumber scenarios ( some before each filter or something ). 

The cucumber features/home_page/map.feature:65  failure I suspect that the cassette is not updated in the step features/step_definitions/basic_steps.rb on line 355.

For the " no implicit conversion of Symbol into Integer (TypeError) " it has to be a problem with the poltergeist driver I will try to fix it in the evening

tansaku commented:
thanks @ZmagoD - I was going to have a little further investigation myself now - I imagine you aren't free to pair :-) I'll update here if I discover anything more

tansaku commented:
digging into " no implicit conversion of Symbol into Integer (TypeError) " it does seem like a poltergeist driver issue - see my debugging output below

```
[227, 236] in /Users/tansaku/.rvm/gems/ruby-2.0.0-p598/gems/poltergeist-1.5.0/lib/capybara/poltergeist/browser.rb
   227:     def cookies
   228:       Hash[command('cookies').map { |cookie| [cookie['name'], Cookie.new(cookie)] }]
   229:     end
   230: 
   231:     def set_cookie(cookie)
=> 232:       if cookie[:expires]
   233:         cookie[:expires] = cookie[:expires].to_i * 1000
   234:       end
   235: 
   236:       command 'set_cookie', cookie
(byebug) cookie
"cookie_policy_accepted=true"
(byebug) cookie[:expires]
```

tansaku commented:
Related SO post with proposed fix: http://stackoverflow.com/questions/17203986/setting-a-cookie-with-javascript-doesnt-work-in-poltergeist-and-or-phantomjs

tansaku commented:
I can make those errors go away by following input from that SO post:

```ruby
Given /^I have a "([^\"]+)" cookie set to "([^\"]+)"$/ do |key, value|
  case Capybara.current_session.driver
    when Capybara::Poltergeist::Driver
      page.driver.set_cookie(key, value)
    when Capybara::RackTest::Driver
      headers = {}
      Rack::Utils.set_cookie_header!(headers,key, value)
      cookie_string = headers['Set-Cookie']
      Capybara.current_session.driver.browser.set_cookie(cookie_string)
    when Capybara::Webkit::Driver
      headers = {}
      Rack::Utils.set_cookie_header!(headers,key, value)
      cookie_string = headers['Set-Cookie']
      Capybara.current_session.driver.browser.set_cookie(cookie_string)
    when Capybara::Selenium::Driver
      page.driver.browser.manage.add_cookie(:name=>key, :value=>value)
    else
      raise "no cookie-setter implemented for driver #{Capybara.current_session.driver.class.name}"
  end
end
```
[note the above hack is a nasty smelly mess and needs serious refactoring before we pull it into the project]

but still get the failing "Then the coordinates for "Age UK" and "Youth UK" should not be the same" and another couple of failures:

```
cucumber features/home_page/map.feature:66 # Scenario: Changing postcode changes the map coordinates
cucumber features/individual_charity_pages/remind_charity_admin_to_update_page.feature:27 # Scenario: Org owner is not reminded to update details prior to a year
cucumber features/individual_charity_pages/this-is-my-org.feature:43 # Scenario: Not Signed in User Who Fails Signin Once
```  

I guess those three will need a detailed look to uncover some issue (maybe timing or some other change) - could be we need to dump the billy cache for them and reload - or something more subtle ...

However I have confirmed to my satisfaction that adding @billy flag to other scenarios does remove all that nasty page.driver output, e.g. bundle exec cucumber features/individual_charity_pages/propose_new_organisation.feature:79  (you probably were already certain of that ...)

Just thinking we might want to enable @billy across the board, or perhaps for every @javascript flag (replace @billy with @javascript perhaps? ), and then disable on a performance basis to get things clean in the first instance ...?

decareano commented:
@tansaku, @zmagod, wanted to post this link for info on Billy, VCR and webmock working together. There are lots of issues there. https://github.com/oesmith/puffing-billy/issues/44. I want to look at the PR tonite. I am at work right now.

ZmagoD commented:
@tansaku @decareano , In the last commit I fixed the no implicit conversion error with @tansaku code ( we will refactor this later firstly I want to fix all the errors ). I had added the "@billy" tag in every scenario that throws a webmock log so now the tests are clean. Now I get 4 errors that online and offline 

offline: 
![screenshot from 2016-01-25 22 02 05](https://cloud.githubusercontent.com/assets/7046787/12564953/6f15b946-c3b2-11e5-9a6f-c2746b451586.png)
online: 
![screenshot from 2016-01-25 22 18 07](https://cloud.githubusercontent.com/assets/7046787/12564958/73f491b2-c3b2-11e5-8696-67bdd5389ee9.png)

tansaku commented:
ran this just now - I get three failing locally while online, but almost no other output from the log, which is really great to see - well done @ZmagoD !

```
(::) failed steps (::)


expected: value != 51.5814584
     got: 51.5814584

(compared using ==)
 (RSpec::Expectations::ExpectationNotMetError)
./features/step_definitions/map_steps.rb:73:in `/^the coordinates for "(.*?)" and "(.*?)" should( not)? be the same/'
features/home_page/map.feature:73:in `Then the coordinates for "Age UK" and "Youth UK" should not be the same'


expected: "/users/sign_in"
     got: "/organisations/1"

(compared using ==)
 (RSpec::Expectations::ExpectationNotMetError)
./features/step_definitions/navigation_steps.rb:50:in `/^I (visit|should be on) the (.*) page$/'
features/individual_charity_pages/this-is-my-org.feature:46:in `Then I should be on the sign in page'

Failing Scenarios:
cucumber features/home_page/map.feature:66 # Scenario: Changing postcode changes the map coordinates
cucumber features/individual_charity_pages/remind_charity_admin_to_update_page.feature:27 # Scenario: Org owner is not reminded to update details prior to a year
cucumber features/individual_charity_pages/this-is-my-org.feature:43 # Scenario: Not Signed in User Who Fails Signin Once

```

immediately I want to clear up one last bit of cruft from the log, since there's a rogue puts statement somewhere :-)

```
$ bundle exec cucumber --format progress
[Coveralls] Using SimpleCov's 'rails' settings.
Using the default profile...
---..............................................................................................................................................................................................................................................................................................................................................................................................---.........I love dogs was not found
I love Cats was found
I hate animals was found
I love people was not found
```

If I have time after this AM standup I'll try and investigate these three failures ...

I see just two failing in the CI

```
cucumber features/home_page/map.feature:66 # Scenario: Changing postcode changes the map coordinates
cucumber features/individual_charity_pages/this-is-my-org.feature:43 # Scenario: Not Signed in User Who Fails Signin Once
254 scenarios (2 failed, 252 passed)
```

makes me think these might well be timing issues, but let's see ...

tansaku commented:
Yeah, timing issues. I just hacked a fix with the following:

```ruby
Given /^I sign in as "(.*?)" with password "(.*?)"( with javascript)?$/ do |email, password, js|
  fill_in "user_email", :with => email
  fill_in "user_password", :with => password
  if js
    page.find("#signin").trigger("click")
    sleep 0.2
  else
    click_link_or_button "Sign in"
  end
end
```

http://stackoverflow.com/questions/24370002/rails-cucumber-how-to-wait-for-javascript-to-finish-before-moving-on-to-next

there are various Capybara methods that will wait for a particular bit of css to load that we should probably prefer to use here, and I got the following to work, which is probably a bit cleaner:

```ruby
Given /^I sign in as "(.*?)" with password "(.*?)"( with javascript)?$/ do |email, password, js|
  fill_in "user_email", :with => email
  fill_in "user_password", :with => password
  if js
    page.find("#signin").trigger("click")
    expect(page).not_to have_css('.new_user')
  else
    click_link_or_button "Sign in"
  end
end
```

tansaku commented:
So the waiting thing clears up two failures, but that lat/long issue is proving a little trickier.

Something about the new setup seems to be preventing the geocoding to be re-triggered in the organisation update, but only in the test suite.  It appears to work fine running the site normally on this branch.

For some reason I can't even hook the debugger on the update flow to check what's going on.  The debugger hooks on fine in development mode.

I've taken the @billy tag on bundle exec cucumber features/home_page/map.feature:66 and it still fails (though passes on develop).  I may be missing something obvious, but feels like there is some subtle issue being introduced here ...

the relevant part of base_organisation.rb is

```
  after_validation :geocode, if: -> { run_geocode? }
  self.table_name = "organisations"

  def run_geocode?
    ## http://api.rubyonrails.org/classes/ActiveModel/Dirty.html
    byebug
    address_changed? or postcode_changed? or (address.present? and not_geocoded?)
  end
```
this byebug is not being triggered when we update the org via the test, even though the operation of the test does appear to correctly update the postcode:

```
[354, 363] in /Users/tansaku/Documents/Github/AgileVentures/LocalSupport/features/step_definitions/basic_steps.rb
   354: 
   355: Given /^I update "(.*?)" charity postcode to be "(.*?)"$/ do |name, postcode|
   356:   steps %Q{And I visit the edit page for the organisation named "#{name}"}
   357:   fill_in('organisation_postcode', :with => postcode)
   358:   byebug
=> 359:   click_button 'Update Organisation'
   360: end
   361: 
   362: Given /^I edit the charity address to be "(.*?)" when Google is indisposed$/ do |address|
   363:   body = %Q({
(byebug) Organisation.find_by name: 'Youth UK'
#<Organisation id: 4, name: "Youth UK", address: "84 pinner road", postcode: "HA1 4HZ", email: "", description: "Care for the Very Young", website: "http://d.com/", telephone: "", created_at: "2016-01-26 10:57:31", updated_at: "2016-01-26 10:57:31", latitude: 51.5814584, longitude: -0.3472856, gmaps: nil, donation_info: "", publish_address: false, publish_phone: false, publish_email: true, deleted_at: nil, type: "Organisation", non_profit: nil>
(byebug) click_button 'Update Organisation'
nil
(byebug) Organisation.find_by name: 'Youth UK'
#<Organisation id: 4, name: "Youth UK", address: "84 pinner road", postcode: "EC4N 4TR", email: "", description: "Care for the Very Young", website: "http://d.com/", telephone: "", created_at: "2016-01-26 10:57:31", updated_at: "2016-01-26 10:57:42", latitude: 51.5814584, longitude: -0.3472856, gmaps: nil, donation_info: "", publish_address: false, publish_phone: false, publish_email: true, deleted_at: nil, type: "Organisation", non_profit: nil>
```

makes me suspicious that byebug is not doing what I expect ... my only strategy going forward would be a step by step mutation from the branch to the develop branch to see which change causes this issue.  There may be other approaches I haven't thought of so I'll leave it for now I guess ... ah, hmm - I think there is something odd going on with the vcr.  That seems to record that  http://maps.googleapis.com/maps/api/geocode/json?address=84%20pinner%20road,%20HA1%204HA&language=en&sensor=false comes back with the same lat/lng, which makes sense if google is auto-correcting the address, but then why do things pass in the develop branch?  Ah, maybe because of the way it is stubbed.

Interesting - I can get the test to pass by hacking up the vcr yml file, but I think what we might have here is that the Google geocoding operation has got smarter, but out tests have been stuck in the past with the hardcoded webmock stuff we had before.  Still just a theory, but fascinating ...

p.s. I think you need to sync upstream/master

tansaku commented:
@ZmagoD thanks for merging develop - let me know when you think you'll be able to work on this next.  Perhaps I should grab it and just put in the fixes I found so we can get this into ASAP?   I'm just thinking it would be great for all devs to be working off this properly sandboxed based ASAP.

But I don't want to step on your toes in case you're really keen to finish this up yourself.   I'll work on other stuff today, but can come back to it tomorrow and work up my own branch if you'd like me to get even more hands on :-)

ZmagoD commented:
Hi @tansaku, if not earlier for the weekend for sure. It will be cool if you put your fixes in this PR.

tansaku commented:
cool @ZmagoD I don't know if I can add fixes directly to this PR since I'm not a collaborator on your LocalSupport repo (you could give me access).  I can however take this branch and pull it into the main LS repo - fix it there and then open a new pull request ...

decareano commented:
@tansaku, @zmagod, I am getting these failures from the latest zmago branch sync with develop. i am only testing the map.feature. https://gist.github.com/decareano/dbd783ff41f2e0eb4e5d

tansaku commented:
@ZmagoD thanks for the access to your repo - I didn't have time to get to this yesterday and I'm off today - will definitely get very involved on Monday

ZmagoD commented:
[354, 363] in /home/zmago/Desktop/AgileVentures/LocalSupport/features/step_definitions/basic_steps.rb
   354: 
   355: Given /^I update "(.*?)" charity postcode to be "(.*?)"$/ do |name, postcode|
   356:   steps %Q{And I visit the edit page for the organisation named "#{name}"}
   357:   fill_in('organisation_postcode', :with => postcode)
   358:   byebug
=> 359:   click_button 'Update Organisation'
   360: end
   361: 
   362: Given /^I edit the charity address to be "(.*?)" when Google is indisposed$/ do |address|
   363:   body = %Q({
(byebug) organisation
*** NameError Exception: undefined local variable or method `organisation' for #<Cucumber::Rails::World:0x00000007a72270>

nil
(byebug) Organisation
Organisation(id: integer, name: string, address: string, postcode: string, email: string, description: text, website: string, telephone: string, created_at: datetime, updated_at: datetime, latitude: float, longitude: float, gmaps: boolean, donation_info: text, publish_address: boolean, publish_phone: boolean, publish_email: boolean, deleted_at: datetime, type: string, non_profit: boolean)
(byebug) Organisation.find_by_name("Youth UK")
#<Organisation id: 4, name: "Youth UK", address: "84 pinner road", postcode: "HA1 4HZ", email: "", description: "Care for the Very Young", website: "http://d.com/", telephone: "", created_at: "2016-01-30 19:14:11", updated_at: "2016-01-30 19:14:11", latitude: 51.5814584, longitude: -0.3472856, gmaps: nil, donation_info: "", publish_address: false, publish_phone: false, publish_email: true, deleted_at: nil, type: "Organisation", non_profit: nil>
(byebug) click_button 'Update Organisation'
{"status"=>"success", "position"=>{"x"=>605.5, "y"=>619}}
(byebug) Organisation.find_by_name("Youth UK")
#<Organisation id: 4, name: "Youth UK", address: "84 pinner road", postcode: "HA1 4HA", email: "", description: "Care for the Very Young", website: "http://d.com/", telephone: "", created_at: "2016-01-30 19:14:11", updated_at: "2016-01-30 19:16:28", latitude: 51.5814584, longitude: -0.3472856, gmaps: nil, donation_info: "", publish_address: false, publish_phone: false, publish_email: true, deleted_at: nil, type: "Organisation", non_profit: nil>

ZmagoD commented:
@decareano I'm getting only these cucumber features/home_page/map.feature:66 failure online and offline now

ZmagoD commented:
@tansaku, @decareano and I have paired on these request and find solution to without_users.feature bug and fixed it. For the features/home_page/map.feature:65 I had put the debuger log in the previous comment. The problem is that the organisation model postcode is get updated but the location is not so we manually change the Changing_postcode_changes_the_map... yml file and get all the scenarios passed online and offline.

tansaku commented:
great work @ZmagoD and @decareano - pairing FTW!

we still do have that single error in the cucumber:

cucumber features/home_page/map.feature:66

is something not checking in perhaps?  I'll check this branch locally ...

tansaku commented:
and I get the same error locally ... double check you didn't miss checking something in and I'll look at this in more detail once I've reviewed some grader issues ...

tansaku commented:
I also notice that I get about 60 new files created in features/req_cache/ which indicates that perhaps some files haven't been checked in ..., e.g.

```
features/req_cache/get_csi.gstatic.com_01a51813f6a5749c22ea4c0e9e989791699168a9.yml
	features/req_cache/get_csi.gstatic.com_02c4e4a36af5eb835a9160bb59d17b36e6458019.yml
	features/req_cache/get_csi.gstatic.com_03152088a5426f4257c98d954b03b0a2384b98df.yml
	features/req_cache/get_csi.gstatic.com_0507fa1c3b5f6b50359854808dfb0d6875f78471.yml
```

ZmagoD commented:
Hi @tansaku Yes every time that I run cucumber the billy's cache is updated. Should we add req_cache dir in gitignore?

tansaku commented:
@ZmagoD we need that to be stable - everything should be checked in - if running cucumber causes the cache to update in a different way then we have a leak of some kind, a problem for us ... we need the sandbox to be stable

tansaku commented:
@ZmagoD as I understand puffing billy it should be allowing us to take tests completely offline, and all interaction data should be fixed in git.  We may have to use some billy options like ignore_params or strip_query_params or something.

The key thing is that we need to work out which cucumber tests are generating new cache hits each time, and find out what's causing it, e.g. is there some randomness in the cucumber test?  Is the google map request including some datetime element in the request so that the cache isn't being used effectively?

Either way, we also have that issue of that one test failing for me locally and on CI, but passing for you.  Please do double check everything is passing locally for you, and double check you have everything checked in.  Push anything remaining and then I'll take a stab at resolving both issues in the week if I can ...

ZmagoD commented:
@tansaku no locally the features/home_page/map.feature:65 still fail

tansaku commented:
Ah sorry @zmagod thought you said you had fixed that locally. You're blocked on that and need me to take a closer look? I'll try and address both issues before the weekend if I can ...

ZmagoD commented:
and the commit: "with the original vcr cache of Changing_postcode_changes_the_map_coor… …
…dinates.yml file" is the last where I restored the original yml files made by puffing-billy. The commit before the "fixed without_users.feature:40 error" we hacked the Changing_postcode_changes_the_map..yml file. We manually change the lat, long properties of that yml and then the tests passed.

tansaku commented:
ah, understood - but so you didn't check in that hack?

Thinking out loud, so we already have a big lot of cached files in features/req_cache, and running the cucumber tests generates more ... I would like to know which cucumber features are generating new hits ... a quick check suggests that some of the new files only differ in terms of the dates.  There are 209 of them - however actually the uris seem unique too, just slight variations, e.g.

http://csi.gstatic.com/csi?v=2&s=mapsapi3&action=map&rt=firsttile.320

vs

http://csi.gstatic.com/csi?v=2&s=mapsapi3&action=map&rt=firsttile.148

I ran the whole test suite offline and everything worked, and no more cache files were generated - so maybe these are just some extra caches that were missed from before and checking them in will sort things ...

interestingly all the new ones return responses like this:

> 'The Google Maps API server rejected your request. The Google Maps JavaScript
  API must be downloaded directly from Google''s servers. Learn more: https://developers.google.com/maps/documentation/javascript/tutorial#Loading_the_Maps_API'

It does look like ignoring params might reduce the number of items in the cache.

However when I run the whole thing again online, I see we have 379 new files.  It feels like something somewhere isn't being cached properly.  We need to fix that ... 

Also we have a failing test for the map feature that is checking that there is some change in geocoding when the postcode changes. However, looking at the responses from the google geocoding service it now looks like the geocoding service is perhaps behaving differently to how it did before, e.g.

* it returns no data if the postcode and address don't match?
* it returns the lat/lng related to the address, independent of the postcode

or some combination of the above.  According to the VCR logs for this scenario,

http://maps.googleapis.com/maps/api/geocode/json?address=84%20pinner%20road,%20HA1%204HA&language=en&sensor=false

gives us the following:


                       "lat" : 51.5814584,
                       "lng" : -0.3472856

while the scenario with the address changes gets the following: http://maps.googleapis.com/maps/api/geocode/json?address=34%20pinner%20road,%20HA1%204HZ&language=en&sensor=false returns the following:

                       "lat" : 51.5814562,
                       "lng" : -0.3439653

Now the VCR log for cucumber in both cases only includes a single request for geocoding.  Other geocoding should take place as part of the background step which it is and that is stored in a separate YAML file specifically for the dummy charity in question, where the request is:

http://maps.googleapis.com/maps/api/geocode/json?address=84%20pinner%20road,%20HA1%204HZ&language=en&sensor=false

and the response is:

                       "lat" : 51.5814584,
                       "lng" : -0.3472856

suggesting that in this case the change of only the postcode leads to no associated change in lat/lng.  Now if I remember the original issue correctly, I think the problem was that we had an organisation with the completely wrong postcode that was showing in the wrong location, and then giving it the correct postcode moved it ...?  So I think for this acceptance test to be doing the right thing it needs to take a charity where there is an address/postcode mismatch followed by setting the postcode correctly ...?

And testing that I think the problem is that  we need to change so that we check the actual lat/lng and that that changes rather than comparing to lat/lng of another, because I think the way the geocoding works has changed slightly.

Hopefully will dig into both the above issues more today ....

Notes:

* the vcr.log should be in .gitignore

tansaku commented:
@ZmagoD I've just adjusted the map feature test to work with the way that geocoding works now - it passes locally - we'll see how the CI does.

That's not the end though, because we really need that cache to be stable.  If it's generating new files on each run then that means that every CI run will still hit 3rd party network services.

Not sure what's going on there - need to drill down on an individual cucumber step ... maybe post a question to the puffing-billy folks ...

tansaku commented:
okay, so running my new `bundle exec cucumber features/home_page/map.feature:70` gives 63 files in the req_cache directory.

When I run it again I have 92 files, so each time we are running it when online we are generating new cache files, which means new network connections with is not good, because it means we are not sandboxed ...

ZmagoD commented:
@tansaku I will try to tweak the billy config in the env.rb if is there some setting like :record_once in VCR

tansaku commented:
@ZmagoD I think I may already have fixed it - it's about the ignore_params setting

tansaku commented:
I've already got it down to 60/61 files - just doing some final tweaks ...

ZmagoD commented:
:+1:

tansaku commented:
hmm, one final add of "http://maps.google.com/maps/vt" tripped a load of:

> One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).

so we've got it so only one new file is added to the cache each time, which is progress, but not quite the cigar.

I'm not sure if we can get away with `js_errors: false` or whether there's some other tweak we need there ...

tansaku commented:
hmm, and that build fails - timing issue probably - gonna have to remove the timing things and replace with `expect(page).not_to have_css('.new_user')` - although that doesn't work locally - rats ...

This is related to the final google maps issue I think:

https://github.com/oesmith/puffing-billy/issues/9

hmmmm

tansaku commented:
alright - we are green and clean - there's a lot of refactoring to do - need to make tickets for that, but perhaps we can pull this in and speed up everyone's test times ...

ZmagoD commented:
Wow @tansaku how do you manage to get rid of features/home_page/map.feature:65 failure?

tansaku commented:
@ZmagoD I rewrote the test so that rather than comparing the new location to an existing orgs location, it compares it with the start location.

Basically the hard coded webmock was doing something that google geocode used to do and so isolating us from a failure that we should have been seeing.  VCR exposed that failure, and now the test makes sense in terms of how google geocode behaves.

If google geocoding behaviour changes again our upgrade path with be simpler - just delete the relevant VCR file.

Notice also that we are down to only 62 files in the req_cache and that now appears to be stable, indicating that we have probably isolated all network connections.

I think there are a couple of other places we need to clean up relating to possible javascript timeouts - and we need to refactor some of the config, but I think we are basically there.  

Will probably merge in after lunch after a further quick review and making the relevant follow up tickets ...

ZmagoD commented:
:+1:

tansaku commented:
added refactoring chore: https://www.pivotaltracker.com/story/show/112899357 and tangential capybara waiting logic issue here: https://www.pivotaltracker.com/story/show/112900047

decareano commented:
@tansaku, @zmagod, in commit `a78b9e9` the tests pass ONLINE but it fails OFFLINE? or they pass both ONLINE and OFFLINE. I tried your workaround for the js_errors but I cannot make it work OFFLINE. I am wandering: is there a way to tell the CI Engine to run tests OFFLINE. is that possible?

ZmagoD commented:
@decareano I get one test fail offline the cucumber features/sign_in_sign_up/sign_in.feature:56 but when I run it again offline it pass. I don't get any js_errors

decareano commented:
@tansaku, even with your latest changes to javascript issues, I still get these .css errors in `features/home_page/map.feature:23`. When I am ONLINE it works but not OFFLINE. Do you have those errors?

@javascript @vcr @billy
  Scenario: Show all charities in map on homepage map          # features/home_page/map.feature:23
    Given I visit the home page                                # features/step_definitions/navigation_steps.rb:45
    Then I should see the following measle markers in the map: # features/step_definitions/map_steps.rb:40
      | Indian Elders Association | Age UK | Harrow Bereavement Counselling |
      expected to find css ".measle" 3 times but there were no matches (RSpec::Expectations::ExpectationNotMetError)
      ./features/step_definitions/map_steps.rb:43:in `/^I should( not)? see the following (measle|vol_op) markers in the map:$/'
      features/home_page/map.feature:25:in `Then I should see the following measle markers in the map:'

Failing Scenarios:
cucumber features/home_page/map.feature:23 # Scenario: Show all charities in map on homepage map

tansaku commented:
@decareano I haven't tried it offline recently ... :-(

tansaku commented:
@decareano just ran entire suite locally OFFLINE, no failures:

```
[tansaku@Samuels-MBP-2:~/Documents/Github/AgileVentures/LocalSupport (develop)]$ 
→ bundle exec rspec -fp
[Coveralls] Using SimpleCov's 'rails' settings.

Randomized with seed 33067
.....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

Top 10 slowest examples (2.95 seconds, 14.6% of total time):
  Moderate a proposed org as non-superadmin does not act upon non superadmins requests to decline/destroy org
    0.46108 seconds ./spec/requests/propose_organisations_spec.rb:14
  AcceptProposedOrganisation proposed organisation email is not registered sends an email
    0.43127 seconds ./spec/services/accept_proposed_organisation_spec.rb:26
  Organisation Creating of Organisations from CSV file must be able to generate multiple Organisations from text file
    0.42593 seconds ./spec/models/organisation_spec.rb:232
  Organisation Creating of Organisations from CSV file must fail gracefully when encountering error in generating multiple Organisations from text file
    0.38946 seconds ./spec/models/organisation_spec.rb:238
  UserOrganisationDecliner#call calls update_message_for_decline_success
    0.29198 seconds ./spec/services/users_organisation_decliner_spec.rb:19
  Organisation Creating of Organisations from CSV file importing category relations must be able to avoid org category relations from text file when org does not exist
    0.21396 seconds ./spec/models/organisation_spec.rb:313
  Organisation Creating of Organisations from CSV file importing category relations must fail gracefully when encountering error in importing categories from text file
    0.20252 seconds ./spec/models/organisation_spec.rb:334
  RegistrationsController POST create redirects to home page after registration form
    0.19595 seconds ./spec/controllers/registrations_controller_spec.rb:20
  Devise::PasswordsController POST create successful redirects to sign-in after user requests password for email in the system
    0.18058 seconds ./spec/controllers/passwords_controller_spec.rb:23
  Queries::Organisations::search_by_keyword_and_category filters by category when searchss by keyword is nil should not include #<Organisation id: 9, name: "Harrow Bereavement Counselling", address: "64 pinner road", postcode: "HA1 4HZ", email: "", description: "Bereavement Counselling", website: "", telephone: "", created_at: "2016-02-04 16:02:05", updated_at: "2016-02-04 16:02:05", latitude: 51.5813411, longitude: -0.3468075, gmaps: nil, donation_info: "http://www.harrow-bereavment.co.uk/donate", publish_address: false, publish_phone: false, publish_email: true, deleted_at: nil, type: "Organisation", non_profit: nil>
    0.15439 seconds ./spec/services/queries/organisations_spec.rb:118

Top 10 slowest example groups:
  Moderate a proposed org
    0.14382 seconds average (0.57526 seconds / 4 examples) ./spec/requests/propose_organisations_spec.rb:3
  Propose an edit
    0.11057 seconds average (0.11057 seconds / 1 example) ./spec/requests/propose_edits_spec.rb:3
  UserOrganisationDecliner#call
    0.08873 seconds average (0.35491 seconds / 4 examples) ./spec/services/users_organisation_decliner_spec.rb:4
  UserReports
    0.08173 seconds average (0.2452 seconds / 3 examples) ./spec/requests/user_reports_spec.rb:3
  Organisation
    0.07673 seconds average (5.14 seconds / 67 examples) ./spec/models/organisation_spec.rb:3
  organisations/_form.html.erb
    0.07537 seconds average (0.30147 seconds / 4 examples) ./spec/views/organisations/form.html.erb_spec.rb:3
  VolunteerOps
    0.07212 seconds average (0.28846 seconds / 4 examples) ./spec/requests/volunteer_ops_spec.rb:3
  Invitations
    0.06494 seconds average (0.12989 seconds / 2 examples) ./spec/requests/invitations_spec.rb:3
  Proposed Edits Moderation
    0.05932 seconds average (0.05932 seconds / 1 example) ./spec/requests/moderate_edits_spec.rb:3
  Devise::PasswordsController
    0.057 seconds average (0.22799 seconds / 4 examples) ./spec/controllers/passwords_controller_spec.rb:3

Finished in 20.2 seconds (files took 5.83 seconds to load)
741 examples, 0 failures

Randomized with seed 33067

[tansaku@Samuels-MBP-2:~/Documents/Github/AgileVentures/LocalSupport (develop)]$ 
→ bundle exec cucumber --format progress
[Coveralls] Using SimpleCov's 'rails' settings.
Using the default profile...
---..............................................................................................................................................................................................................................................................................................................................................................................................---.........I love dogs was not found
I love Cats was found
I hate animals was found
I love people was not found
..........-.........................................WARNING: The next major version of capybara-webkit will require at least version 5.0 of Qt. You're using version 4.8.6.
[DEPRECATION] Capybara::Webkit::Driver#browser is deprecated.
........................................................................................................................---............................................................................................................................................---.........................................................................................................................----.........................................---.......................................................................................................................-...--.........................................................--------................................................................................................---...........................---..................-------...........................................................................................................................................................................................................................................................................................................----------.............................................................................................................................................................---....---...............................................................................................--.---.---..........---........................

254 scenarios (254 passed)
2018 steps (2018 passed)
1m20.529s
```

check your git status - do you have latest develop and no hanging around files?

Check out a fresh version to be sure maybe?  We have a lot of things in .gitignore ...

decareano commented:
@tansaku, my rspec is sandboxed pretty well. No errors OFFLINE. I was referring to Cucumber tests in my comment above.

tansaku commented:
@decareano look closely at what I posted - it includes both rspec AND cucumber

decareano commented:
@tansaku, I see what you are saying. So, develop now has the latest and greatest? did you merge it?

tansaku commented:
@decareano sorry - I often miss questions in closed pull requests - I think everything is merged ... are you still having trouble?

I am still seeing occasional js failures ....

decareano commented:
@tansaku, all is well. No worries.
